src/packagedcode/go.py

```diff
@@ -0,0 +1,256 @@
+import re
+
+import attr
+from packageurl import PackageURL
+
+from commoncode import filetype
+from commoncode import fileutils
+from packagedcode import models
+
+
+"""
+Handle Go packages
+"""
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GoPackage(models.Package):
+    metafiles = ('go.mod',)
+    default_type = 'gopkg'
+    default_primary_language = 'Go'
+    default_web_baseurl = None
+    default_download_baseurl = None
+    default_api_baseurl = None
+
+    @classmethod
+    def recognize(cls, location):
+        if fileutils.file_name(location).lower() == 'go.mod':
+            gomod_data = parse_gomod(location)
+            yield build_gomod_package(gomod_data)
+
+    @classmethod
+    def get_package_root(cls, manifest_resource, codebase):
+        return manifest_resource.parent(codebase)
+
+    def repository_homepage_url(self):
+        return self.homepage_url
+
+    def repository_download_url(self):
+        return self.download_url
+
+
+def build_gomod_package(gomod_data):
+    """
+    Return a Pacakge object from a package data mapping or None.
+    """
+    name = gomod_data.get('name')
+    homepage_url = "https://{}".format(gomod_data.get('module'))  
+    download_url = "https://{}/archive/master.zip".format(gomod_data.get('module')) 
+
+    package_dependencies = []
+    require = gomod_data.get('require')
+    if require:
+        for name, version in require:
+            package_dependencies.append(
+                models.DependentPackage(
+                    purl=PackageURL(
+                        type='gopkg',
+                        name=name,
+                        version=version
+                    ).to_string(),
+                    requirement=version,
+                    scope='require',
+                    is_runtime=True,
+                    is_optional=False,
+                    is_resolved=True,
+                )
+            )
+    exclude = gomod_data.get('exclude')
+    if exclude:
+        for name, version in exclude:
+            package_dependencies.append(
+                models.DependentPackage(
+                    purl=PackageURL(
+                        type='gopkg',
+                        name=name,
+                        version=version
+                    ).to_string(),
+                    requirement=version,
+                    scope='exclude',
+                    is_runtime=True,
+                    is_optional=False,
+                    is_resolved=True,
+                )
+            )
+
+    return GoPackage(
+        name=name,
+        homepage_url=homepage_url,
+        download_url=download_url,
+        code_view_url=homepage_url,
+        dependencies=package_dependencies
+    )
+
+
+####################################################################
+# Regrex expressions to parse different types of dependency
```

Comment:
I think Regrex is a typo
Also please put an example before each expression for illustration

Reply:
updated my PR :)